Source Google-Ads: move to next releaseStage

tolik0 opened the pull request:
<!--
Thanks for your contribution! 
Before you submit the pull request, 
I'd like to kindly remind you to take a moment and read through our guidelines
to ensure that your contribution aligns with the type of contributions our project accepts.
All the information you need can be found here:
   https://docs.airbyte.com/contributing-to-airbyte/

We truly appreciate your interest in contributing to Airbyte,
and we're excited to see what you have to offer! 

If you have any questions or need any assistance, feel free to reach out in #contributions Slack channel.
-->

## What
Resolve:
* #28204 

## How
Delete the required parameter - start date. Add missing fields in schemas. Add primary keys to streams.

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
*Are there any breaking changes? What is the end result perceived by the user?*

sherifnada commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:
```diff
@@ -460,12 +486,172 @@ class Audience(GoogleAdsStream):
+        return query
+
+
+class IncrementalEventsStream(GoogleAdsStream, IncrementalMixin, ABC):
```
can you explain conceptually how the `IncrementalEventsStream` works and relates to other streams? How does this replicate deletes exactly? I'm having a hard time understanding how this achieves delete replication. Unit testing would also help me understand how this works.

tolik0 replied:
I added unit tests and an explanation of IncrementalEventsStream to the docs.

lazebnyi replied:
Please add docs for this class

tolik0 replied:
Added docs

sherifnada commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:
```diff
@@ -460,12 +486,172 @@ class Audience(GoogleAdsStream):
+        record_fields = GoogleAds.get_fields_from_schema(self.get_json_schema())
+        deleted_record = {field_: None for field_ in record_fields}
+        for id_ in stream_slice.get("deleted_ids", []):
+            deleted_record[self.id_field] = id_
```
`self.id_field` is set to `None` above. Is it ever reset to something else?

sherifnada replied:
I see that it's used below. This is a confusing way of declaring those instance variables. They should just be set in the constructor and not allowed to be `None`

tolik0 replied:
Fixed

sherifnada commented on airbyte-integrations/connectors/source-google-ads/unit_tests/test_source.py:
```diff
@@ -145,7 +145,7 @@ def test_chunk_date_range():
 def test_streams_count(config, mock_account_info):
```
there should be unit tests for the newly added streams

tolik0 replied:
Done

sherifnada reviewed (changes requested):
Hi @tolik0 👋🏼 I left inline comments about the PKs and FKs. But my main outstanding question is how `IncrementalEvents` are supposed to work

tolik0 commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/schemas/ad_group_ads.json:
```diff
@@ -2,6 +2,9 @@
   "$schema": "http://json-schema.org/draft-07/schema#",
   "type": "object",
   "properties": {
+    "customer.id": {
```
I don't find any documentation that states the uniqueness of IDs between different accounts, so customer ID should be added to all streams where it can be requested. For other resources, I will add parent resource id to the stream schema.

tolik0 replied:
Will be fixed in different pr: https://github.com/airbytehq/airbyte/pull/30091

tolik0 commented on airbyte-integrations/connectors/source-google-ads/acceptance-test-config.yml:
```diff
@@ -20,32 +20,168 @@ acceptance_tests:
+        timeout_seconds: 600
+        empty_streams:
+          - name: "account_performance_report"
+            bypass_reason: "Floating data"
```
Added better-explained reason

tolik0 commented on airbyte-integrations/connectors/source-google-ads/acceptance-test-config.yml:
```diff
@@ -20,32 +20,168 @@ acceptance_tests:
+            bypass_reason: "Floating data"
           - name: "accounts"
             bypass_reason: "Floating data"
+          - name: "ad_group_ad_labels"
```
Done

tolik0 commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:
```diff
@@ -460,12 +486,172 @@ class Audience(GoogleAdsStream):
     """
 
-    primary_key = ["audience.id"]
+    primary_key = ["customer.id", "audience.id"]
```
Fixed

tolik0 commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py:
```diff
@@ -80,14 +89,18 @@ def get_credentials(config: Mapping[str, Any]) -> MutableMapping[str, Any]:
 
     @staticmethod
     def get_incremental_stream_config(google_api: GoogleAds, config: Mapping[str, Any], customers: List[Customer]):
+        start_date = config.get("start_date")
```
Fixed in different pr: https://github.com/airbytehq/airbyte/pull/30071

tolik0 commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py:
```diff
@@ -152,12 +171,15 @@ def streams(self, config: Mapping[str, Any]) -> List[Stream]:
+        events_stream = ChangeStatus(api=google_api, customers=customers)
+        incremental_events_config = {"api": google_api, "customers": customers, "parent_stream": events_stream}
         streams = [
+            events_stream,
```
Deleted events from available streams

tolik0 commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:
```diff
@@ -460,12 +486,172 @@ class Audience(GoogleAdsStream):
+            to_date=stream_slice.get("end_date"),
+            cursor_field=self.cursor_field,
+            resource_type=stream_slice.get("resource_type"),
+            limit=10000,
```
Yes, we will lose data, but range_days parameter should be decreased if there are more updates in two weeks. But the limit is a mandatory parameter for this query, and 10k is the maximum possible value.

tolik0 commented on airbyte-integrations/connectors/source-google-ads/acceptance-test-config.yml:
```diff
@@ -20,32 +20,168 @@ acceptance_tests:
+        timeout_seconds: 3600
+        empty_streams:
+          - name: "account_labels"
+            bypass_reason: "Unsufficient permissions"
```
I already asked Kat Wilson to increase permissions for our developer tokens to get data for this stream

lazebnyi commented on airbyte-integrations/connectors/source-google-ads/acceptance-test-config.yml:
```diff
@@ -20,39 +19,53 @@ acceptance_tests:
+            bypass_reason: "Unsufficient permissions"
           - name: "unhappytable"
-            bypass_reason: "Stream not filled yet."
+            bypass_reason: "Empty testing stream"
```
can we populate this stream with data?

tolik0 replied:
Purposely empty stream; that's why it is called unhappytable

tolik0 replied:
Stream was deleted

lazebnyi commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py:
```diff
@@ -120,16 +122,46 @@ def get_fields_from_schema(schema: Mapping[str, Any]) -> List[str]:
         from_category = REPORT_MAPPING[report_name]
         fields = GoogleAds.get_fields_from_schema(schema)
+        if report_name != "change_status" and "change_status.last_change_date_time" in fields:
+            fields.remove("change_status.last_change_date_time")
```
please add comment why we need remove `last_change_date_time` field

tolik0 replied:
Added

lazebnyi commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py:
```diff
@@ -120,16 +122,46 @@ def get_fields_from_schema(schema: Mapping[str, Any]) -> List[str]:
+            id_list_str = ", ".join(f"'{str(id_)}'" for id_ in id_list)
+            conditions.append(f"{id_field} IN ({id_list_str})")
+
+        if resource_type:
```
please add comment like in line 149 with condition description

tolik0 replied:
Added

lazebnyi commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/schemas/accounts.json:
```diff
@@ -42,7 +42,7 @@
     },
     "customer.optimization_score_weight": {
-      "type": ["null", "number"]
+      "type": ["null", "number", "integer"]
```
I believe we change this in other PR

tolik0 replied:
Fixed

lazebnyi commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/schemas/ad_group_criterions.json:
```diff
@@ -186,7 +192,7 @@
     },
     "ad_group_criterion.url_custom_parameters": {
-      "type": ["null", "array"],
+      "type": ["null", "string", "array"],
```
we could not support two data type for one field, better validate type and update it to one if needed

lazebnyi commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py:
```diff
@@ -157,19 +166,22 @@ def streams(self, config: Mapping[str, Any]) -> List[Stream]:
         incremental_config = self.get_incremental_stream_config(google_api, config, customers)
         non_manager_incremental_config = self.get_incremental_stream_config(google_api, config, non_manager_accounts)
+        events_stream = ChangeStatus(api=google_api, customers=customers)
+        incremental_events_config = {"api": google_api, "customers": customers, "parent_stream": events_stream}
```
should be part of `get_incremental_stream_config`

lazebnyi commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:
```diff
@@ -472,3 +464,187 @@ class Labels(GoogleAdsStream):
+        for id_ in stream_slice.get("deleted_ids", []):
+            deleted_record = deleted_record_template.copy()
+            deleted_record[self.id_field] = id_
+            if deleted_record.get(self.primary_key[0]) in stream_slice.get("id_to_time"):
+                deleted_record[self.cursor_field] = stream_slice["id_to_time"][deleted_record[self.primary_key[0]]]
+            else:
+                deleted_record[self.cursor_field] = None
```
```suggestion
            deleted_record[self.cursor_field] = None
            
            slice_exist_for_customer_id = deleted_record.get(self.primary_key[0]) in stream_slice.get("id_to_time")
            
            if slice_exist_for_customer_id:
                deleted_record[self.cursor_field] = stream_slice["id_to_time"][deleted_record[self.primary_key[0]]]
```

sherifnada commented:
@tolik0 removed review request from me & extensibiility -- please reassign once internal GL review is done

tolik0 commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:
```diff
@@ -472,3 +464,194 @@ class Labels(GoogleAdsStream):
+        for id_ in stream_slice.get("deleted_ids", []):
+            deleted_record = deleted_record_template.copy()
+            deleted_record[self.id_field] = id_
+            deleted_record[self.cursor_field] = stream_slice["id_to_time"].get(deleted_record[self.primary_key[0]])
```
Yes, it is always present

lazebnyi commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py:
```diff
@@ -120,16 +115,48 @@ def get_fields_from_schema(schema: Mapping[str, Any]) -> List[str]:
     ) -> str:
         from_category = REPORT_MAPPING[report_name]
         fields = GoogleAds.get_fields_from_schema(schema)
+        # field is a cursor for successors of IncrementalEventsStream, so should be present in schema, but it is obtained from parent stream
```
```suggestion
        
        # field is a cursor for successors of IncrementalEventsStream, so should be present in schema, but it is obtained from parent stream
```

lazebnyi commented on airbyte-integrations/connectors/source-google-ads/acceptance-test-config.yml:
```diff
@@ -20,18 +20,35 @@ acceptance_tests:
-            bypass_reason: "Stream not filled yet."
           - name: "account_labels"
-            bypass_reason: "Unable to seed the stream"
+            bypass_reason: "Unsufficient permissions"
```
Please include the specific permissions we're lacking as the reason for bypassing the testing of this stream.

lazebnyi reviewed (approved):
Need update bypass reason and fix minor comment. Other LGTM!

sherifnada reviewed (commented):
* All [the resource types](https://developers.google.com/google-ads/api/docs/change-event) which are supported by `ChangeStatus` should be impacted by this, no? 
* From a user point of view, it's not an acceptable outcome to say that there will be data loss. If 10k is the maximum then we'll need to find a way around that e.g: make a request for data after date `t` and if it's 10k records then we make another request with date = `max(records_returned_in_first_request)` and so on and so forth

sherifnada commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:
```diff
@@ -472,3 +464,196 @@ class Labels(GoogleAdsStream):
+        return query
+
+
+class IncrementalEventsStream(GoogleAdsStream, IncrementalMixin, ABC):
```
this should follow the same pattern for deletions as shopify, could you sync with @bazarnov about how to do that?

tolik0 replied:
@sherifnada No, not all resources from Change Status should be affected by the new class.
We already fetch incremental updates for resources like Campaigns, Ad Groups, Ad Group Ads, and Campaign Budget based on date.
Also, these resources, unlike criterions, can't be deleted, only marked as "Removed". Our approach captures this.
Given these points, using ChangeStatus for all resources isn't necessary. Our method ensures we capture all relevant updates efficiently.

sherifnada replied:
Great, thanks for the explanation! Can you add this context to the docstring for this class? For the next developer reading this, it wouldn't be obvious why we're using this mechanism for two streams only so it is helpful to retain this knowledge

bazarnov commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:
```diff
@@ -472,3 +515,189 @@ class Labels(GoogleAdsStream):
+    """
+
+    cursor_field = "change_status.last_change_date_time"
+    slice_duration = pendulum.duration(days=15)
```
Is there a reason for having a 14-day period for `IncrementalGoogleAdsStream ` and replace it with 15-day period here?

tolik0 replied:
Fixed

bazarnov reviewed (approved):
Looks good to me, with all the nuances of Google Ads API.

sherifnada commented:
@tolik0 could you address the point about the 10k limit above? 

> From a user point of view, it's not an acceptable outcome to say that there will be data loss. If 10k is the maximum then we'll need to find a way around that e.g: make a request for data after date t and if it's 10k records then we make another request with date = max(records_returned_in_first_request) and so on and so forth

tolik0 commented:
@sherifnada I fixed 10k records for a day by decreasing the granularity of the cursor to microseconds. I also updated the corresponding documentation. 

> @tolik0 could you address the point about the 10k limit above?
> 
> > From a user point of view, it's not an acceptable outcome to say that there will be data loss. If 10k is the maximum then we'll need to find a way around that e.g: make a request for data after date t and if it's 10k records then we make another request with date = max(records_returned_in_first_request) and so on and so forth

katmarkham commented:
Does this support : #11539  ?

sherifnada commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:
```diff
@@ -472,3 +528,207 @@ class Labels(GoogleAdsStream):
+    """
+
+    def __init__(self, parent_stream: IncrementalGoogleAdsStream, **kwargs):
+        self.parent_stream = parent_stream
```
why is parent stream parametrized? isn't it always `ChangeStatus`?

tolik0 replied:
Fixed

sherifnada commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:
```diff
@@ -472,3 +528,207 @@ class Labels(GoogleAdsStream):
+        return query
+
+
+class AdGroupCriterions(IncrementalEventsStream):
```
we should be consistent with plural vs. singular in naming, this should be called `*Criterion` not `*Criterions`

tolik0 replied:
This is the old name. To avoid blocking this PR and certification, I created [new issue](https://github.com/airbytehq/airbyte/issues/30562) for this change as it is a breaking change.

sherifnada commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:
```diff
@@ -472,3 +528,207 @@ class Labels(GoogleAdsStream):
+    cursor_field = "change_status.last_change_date_time"
+
+
+class AdListingGroupCriterions(AdGroupCriterions):
```
what is the difference between this class and `AdGroupCriterion`? the docstring is the same

tolik0 replied:
These streams have different configurations of schemas, AdListingGroupCriterion request fields about listing_group.

sherifnada replied:
ok, can you list that in the docstring with the appropraite URL?

tolik0 replied:
Added

sherifnada commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:
```diff
@@ -472,3 +528,207 @@ class Labels(GoogleAdsStream):
+        self, sync_mode: SyncMode, cursor_field: Optional[str], stream_state: Mapping[str, Any]
+    ) -> Iterable[Mapping[str, Any]]:
+        for parent_slice in self.parent_stream.stream_slices(
+            sync_mode=sync_mode, cursor_field=cursor_field, stream_state=stream_state.get(self.parent_stream_name)
```
is the sync mode ever not incremental? It was confusing to have to scroll up and down through multiple levels of indirection to see that it's only ever set to incremental, I think it would be easier to read if incremental was just hardcoded here

tolik0 replied:
Fixed

sherifnada commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:
```diff
@@ -472,3 +528,207 @@ class Labels(GoogleAdsStream):
+            slices_generator = self.read_parent_stream(self.parent_sync_mode, self.parent_cursor_field, stream_state)
+            yield from slices_generator
+        else:
+            for customer in self.customers:
```
why are we keeping the IDs of all deleted records in the stream slice, instead of doing something like this in `read_records` : 

```python
for record in parent.read_records():
  if parent_record.get("change_status.resource_status") == "REMOVED":
    yield {self.id_field: parent_record.get(id), "deleted_at": parent_record.get(updated_at)}
```
? 

It feels like structuring this this way inlined in `read_records` would be a lot simpler to understand

tolik0 replied:
During stream slicing, we collect ids of updated records and then request them in one batch. To avoid reading parent two times during slicing and reading deleting records we save results in slicing.

sherifnada commented on airbyte-integrations/connectors/source-google-ads/unit_tests/test_incremental_events_streams.py:
```diff
@@ -0,0 +1,124 @@
+    result = list(
+        stream.read_records(sync_mode=SyncMode.incremental, cursor_field=["change_status.last_change_date_time"],
+                            stream_slice=stream_slice))
+    assert len(result) == 4
```
can you assert on the content of the records too?

tolik0 replied:
Fixed

tolik0 commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:
```diff
@@ -196,53 +294,21 @@ def read_records(
-                # return the control if no exception is raised
-                self.incremental_sieve_logger.info("Current slice has been read. Exiting read_records()")
-                return
+                if not self._handle_records_limit_hit(records_count, stream_slice, customer_id):
```
Fixed

tolik0 commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:
```diff
@@ -89,17 +113,18 @@ def __init__(self, api: GoogleAds, customers: List[Customer]):
         return query
 
-    def parse_response(self, response: SearchPager) -> Iterable[Mapping]:
+    def parse_response(self, response: SearchPager, stream_slice: Optional[Mapping[str, Any]] = None) -> Iterable[Mapping]:
```
It is used in IncrementalEventsStream

tolik0 commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py:
```diff
@@ -113,16 +115,53 @@ def get_fields_from_schema(schema: Mapping[str, Any]) -> List[str]:
+
+        # filter ChangeStatus events by type of resource
+        if resource_type:
+            conditions.append(f"change_status.resource_type = '{resource_type}'")
```
Fixed

tolik0 commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py:
```diff
@@ -113,16 +115,53 @@ def get_fields_from_schema(schema: Mapping[str, Any]) -> List[str]:
+        # field is a cursor for successors of IncrementalEventsStream, so should be present in schema, but it is obtained from parent stream
+        if report_name != "change_status" and "change_status.last_change_date_time" in fields:
+            fields.remove("change_status.last_change_date_time")
+        # deleted_at used in IncrementalEventsStream for deleted records, it is obtained from stream ChangeStatus
```
Fixed

tolik0 commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/schemas/ad_group_criterions.json:
```diff
@@ -2,6 +2,15 @@
+    "change_status.last_change_date_time": {
+      "type": ["null", "string"]
+    },
+    "customer.id": {
```
Fixed

tolik0 commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/schemas/campaign_criterion.json:
```diff
@@ -0,0 +1,39 @@
+    "change_status.last_change_date_time": {
+      "type": ["null", "string"]
+    },
+    "customer.id": {
```
Fixed

sherifnada commented on airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:
```diff
@@ -472,3 +466,261 @@ class Labels(GoogleAdsStream):
+                            message=f"More then limit {self.mandatory_limit} records with same cursor field. Incremental sync is not possible for this stream.",
+                            failure_type=FailureType.system_error,
+                        )
+                return
```
shouldn't this return be inside an `else`? Otherwise we'd update the stream slice then return from the method anyways. 

Could you test this behavior in a unit test?

lazebnyi reviewed (approved):
LGTM
